### DISCUSSION TITLE
Run queries as authenticated user with direct connection

### DISCUSSION THREAD

**Discussion by chmoder** [supabase/supabase#30124] [Questions] [answered]
https://supabase.com/docs/guides/functions/connect-to-postgres#using-a-postgres-client

Hi, this guide shows how to connect directly to Postgres in an edge function as the Postgres user, but not as the user authenticated with the JWT. 

Any advice on how to connect or run queries with that user would help secure my service.

**Comment by GaryAustin1** [comment, marked as answer]
Basically you have to set a Postgres config variable (request.jwt.claims) to the jwt claims from the user you need for RLS (for sure sub UUID claim and likely role claim).  Then you have to run the SQL as `authenticated` Postgres role.

This is one example I found quickly.   I've seen better ones but can't find them.
https://github.com/orgs/supabase/discussions/7000

**Reply by chmoder** [reply]
Thank you!   This got me on the right track.

```TypeScript
  function enforceRlsLocal() {
    return `
      BEGIN;

      -- Set JWT claims
      DO $$
      BEGIN
        PERFORM set_config(
          'request.jwt.claims',
          jsonb_build_object(
            'role', 'authenticated',
            'sub', '${user.id}'
          )::text,
          true
        );
      END $$;

      -- Set the PostgreSQL role
      DO $$
      BEGIN
        PERFORM set_config('role', 'authenticated', true);
      END $$;

      COMMIT;
  `;
  }

  function enforceRlsSession() {
    return `
      set session request.jwt.claims to '{"role": "authenticated", "sub": "${user.id}"}';
      set session role to 'authenticated';
  `;
  }
```

I tried using `enforceRlsLocal` with the desired query in the transaction but for some reason this messes up langchain.  So then I tried calling it without the query embedded as shown, then later executing the query.  That doesn't apply rls either (because it is a `LOCAL` change.  So now I am using `enforceRlsSession`, which works but sets the session.

Should I add [RESET](https://www.postgresql.org/docs/current/sql-reset.html) before the HTTP request ends?

**Reply by GaryAustin1** [reply]
The variables should get set for each transaction/session from http by PostgREST.  But this is not something I've used for anything but debugging RLS from an IDE.

**Reply by chmoder** [reply]
This will work for us.  For anyone coming across this thread in the future, you can test to see if RLS is working by running the SQL statement `SELECT auth.uid()`.

Thanks for the help!

**Reply by sheariley** [reply]
@chmoder You should use parameterized queries instead of injecting values. Otherwise you are susceptible to SQL injection attacks.

**Reply by chmoder** [reply]
@sheariley,

Thanks for the heads-up!  I am so glad you said something...  I was trying so hard to "just get this to work" that I didn't even think about basic security.

We refactored our RLS helpers so we never interpolate user ids into SQL anymore. Both the Deno edge functions and the Node `pg` client now set `request.jwt.claims` and the `authenticated` role via parameterized `set_config` calls, so the payload is always JSON-stringified before it hits Postgres. Example from our shared helper:

```39:42:platform/supabase/functions/_shared/rls.ts
export async function enforceRlsSession(client: PgClient, userId: string) {
  await setJwtClaims(client, userId, false);
  await setRole(client, false);
}
```

…and the helper itself binds everything:

```10:31:platform/supabase/functions/_shared/rls.ts
await client.queryArray(
  'select set_config($1, $2, $3)',
  [key, value, isLocal],
);
```

If there is anything I can do for you, just let me know.

**Reply by sheariley** [reply]
Nicely done. 👍

**Reply by sheariley** [reply]
Just in case someone with the same situation as me comes across this:

Preface: I know this isn't the best way for this stack to be used to implement this, but I started this project to learn several things and refresh my memory on others.

My situation:
* I'm using the Supabase CLI to setup, configure, and migrate the DB in my dev environment.
* I'm connecting to the database through the session POOLER from a .NET GraphQL API; using Entity Framework.
* The .NET GraphQL API validates the JWT and then sets the claims and role as stated in the accepted answer, with the exception that I had to set the 3rd argument of the `set_config` function to `false`. Otherwise, it would still bypass the RLS on my tables.

Here is my `DbConnectionInterceptor` implementation (registered via `builder.Services.AddDbContext`):
``` csharp
  public class JwtSessionInterceptor : DbConnectionInterceptor
  {
    private readonly IHttpContextAccessor _httpContextAccessor;
    private readonly IUserIdAccessor _userIdAccessor;

    public JwtSessionInterceptor(
      IHttpContextAccessor accessor,
      IUserIdAccessor userIdAccessor
    )
    {
      _httpContextAccessor = accessor;
      _userIdAccessor = userIdAccessor;
    }

    public override async Task ConnectionOpenedAsync(
      DbConnection connection,
      ConnectionEndEventData eventData,
      CancellationToken cancellationToken = default
    )
    {
      var context = _httpContextAccessor.HttpContext;
      var userId = _userIdAccessor.GetUserId();

      var isAuthenticated = context?.User?.Identity?.IsAuthenticated == true && userId != null;
      var dbRole = isAuthenticated ? "authenticated" : "anon";
      
      var claims = new Dictionary<string, string> { { "role", dbRole } };
      if (isAuthenticated) claims.Add("sub", userId.ToString()!);
      var claimsJson = JsonSerializer.Serialize(claims);

      using var cmd = connection.CreateCommand();
      cmd.CommandText = """
        SELECT set_config('request.jwt.claims', @claims, false);
        SELECT set_config('role', @role, false);
      """;
      
      var claimsParam = cmd.CreateParameter();
      claimsParam.ParameterName = "claims";
      claimsParam.Value = claimsJson;
      cmd.Parameters.Add(claimsParam);

      var roleParam = cmd.CreateParameter();
      roleParam.ParameterName = "role";
      roleParam.Value = dbRole;
      cmd.Parameters.Add(roleParam);

      await cmd.ExecuteNonQueryAsync(cancellationToken);
    }
  }
```
